snappy/desktop.go

```diff
@@ -0,0 +1,164 @@
+	regexp.MustCompile(`^\[Desktop Action [a-zA-Z0-9]+\]$`),
+	// whitespace lines
+	regexp.MustCompile(`^\s*$`),
+	// lines with comments
+	regexp.MustCompile(`^\s*#`),
+	// https://specifications.freedesktop.org/desktop-entry-spec/latest/ar01s05.html
+	regexp.MustCompile(`^Type=`),
+	regexp.MustCompile(`^Version=`),
+	regexp.MustCompile(fmt.Sprintf(`^Name%s=`, desktopFileI18nPattern)),
+	regexp.MustCompile(fmt.Sprintf(`^GenericName%s=`, desktopFileI18nPattern)),
+	regexp.MustCompile(`^NoDisplay=`),
+	regexp.MustCompile(fmt.Sprintf(`^Comment%s=`, desktopFileI18nPattern)),
+	regexp.MustCompile(`^Icon=`),
+	regexp.MustCompile(`^Hidden=`),
+	regexp.MustCompile(`^OnlyShowIn=`),
+	regexp.MustCompile(`^NotShowIn=`),
+	regexp.MustCompile(`^Exec=`),
+	// Note that we do not support TryExec, it does not make sense
+	// in the snap context
+	regexp.MustCompile(`^Terminal=`),
+	regexp.MustCompile(`^Actions=`),
+	regexp.MustCompile(`^MimeType=`),
+	regexp.MustCompile(`^Categories=`),
+	regexp.MustCompile(fmt.Sprintf(`^Keywords%s=`, desktopFileI18nPattern)),
+	regexp.MustCompile(`^StartupNotify=`),
+	regexp.MustCompile(`^StartupWMClass`),
+}
+
+func isValidDesktopFilePrefix(line string) bool {
+	for _, re := range validDesktopFileLines {
+		if re.MatchString(line) {
+			return true
+		}
+	}
+	return false
+}
+
+func rewriteExecLine(m *snapYaml, line string) (string, error) {
+	cmd := strings.SplitN(line, "=", 2)[1]
+	for _, app := range m.Apps {
+		validCmd := filepath.Base(generateBinaryName(m, app))
+		// check the prefix to allow %flag style args
+		// this is ok because desktop files are not run through sh
+		// so we don't have to worry about the arguments too much
+		if cmd == validCmd || strings.HasPrefix(cmd, validCmd+" ") {
+			binDir := stripGlobalRootDir(dirs.SnapBinariesDir)
+			absoluteCmd := filepath.Join(binDir, cmd)
+			return strings.Replace(line, cmd, absoluteCmd, 1), nil
+		}
+	}
+
+	return "", fmt.Errorf("invalid exec command: %q", cmd)
+}
+
+func sanitizeDesktopFile(m *snapYaml, realBaseDir string, rawcontent []byte) []byte {
+	newContent := []string{}
+
+	scanner := bufio.NewScanner(bytes.NewReader(rawcontent))
+	for scanner.Scan() {
+		line := scanner.Text()
+		// ignore everything we have not whitelisted
+		if !isValidDesktopFilePrefix(line) {
+			continue
+		}
+		// rewrite exec lines to an absolute path for the binary
+		if strings.HasPrefix(line, "Exec=") {
+			var err error
+			line, err = rewriteExecLine(m, line)
+			if err != nil {
+				// something went wrong, ignore the line
+				continue
+			}
+		}
+
+		// do variable substitution
+		line = strings.Replace(line, "${SNAP}", realBaseDir, -1)
+		newContent = append(newContent, line)
+	}
+
+	return []byte(strings.Join(newContent, "\n"))
+}
+
+func addPackageDesktopFiles(m *snapYaml, baseDir string) error {
+	if err := os.MkdirAll(dirs.SnapDesktopFilesDir, 0755); err != nil {
+		return err
+	}
+
+	desktopFiles, err := filepath.Glob(filepath.Join(baseDir, "meta", "gui", "*.desktop"))
```

Comment:
Thanks! So far I was not considering restricting the name of the desktop files. Should we do this in order to help unity/BAMF to match desktop-file to application? It feels a little heavy handed to restrict it so that application `foo` can only have `meta/gui/foo.desktop` but if it is actually important for unity/BAMF I can change that of course.